### ISSUE TITLE
TPS recording incorrectly

### ISSUE THREAD

**Issue by SlimeDog** [Rsl1122/Plan-PlayerAnalytics#524]
Spigot 1.12.2
Plan 4.1.5

Server **/tps** and **/toplite:top** show consistent 19.6x for 1 min, 5 min, and 15 min. Plan is recording 20.0 and 19.99 in the database every minute.

**Comment by SlimeDog** [comment]
Just to confirm:

After recent plugin updates (unrelated to Plan), my average TPS has dropped from 19.9 to 19.6. (I am trying to isolate that issue.) But Plan 4.1.5 still records 20.0 consistently. So a more accurate method is needed.

**Comment by Rsl1122** [comment]
I went ahead and checked my math.

| TPS | ms / Tick | 20 ticks ms (Diff) |
|-----|-------------|--------------------|
| 20 | 50 | 1000 |
| 19 | 52,63157895 | 1052,631579 |
| 18 | 55,55555556 | 1111,111111 |
| 17 | 58,82352941 | 1176,470588 |
| 16 | 62,5 | 1250 |
| 15 | 66,66666667 | 1333,333333 |
| 14 | 71,42857143 | 1428,571429 |
| 13 | 76,92307692 | 1538,461538 |
| 12 | 83,33333333 | 1666,666667 |
| 11 | 90,90909091 | 1818,181818 |
| 10 | 100 | 2000 |
| 9 | 111,1111111 | 2222,222222 |
| 8 | 125 | 2500 |
| 7 | 142,8571429 | 2857,142857 |
| 6 | 166,6666667 | 3333,333333 |
| 5 | 200 | 4000 |
| 4 | 250 | 5000 |
| 3 | 333,3333333 | 6666,666667 |
| 2 | 500 | 10000 |
| 1 | 1000 | 20000 |

The numbers above are just some values that could occur.
The tables calculate the following formula 20000 / (1000 / (1000 / TPS)) = Diff
which reduces to
TPS = 20000 / Diff

In the code:
```
long difference = diff;
        /* This if braket eliminates tps above 20, when diff is less than 1000 ms (see table) */
        if (difference < TimeAmount.SECOND.ns()) {
            difference = TimeAmount.SECOND.ns();
        }

        long twentySeconds = 20L * TimeAmount.SECOND.ns();
        while (difference > twentySeconds) {
            // Add 0 TPS since more than a second has passed
            history.add(new TPS(now, 0, playersOnline, cpuUsage, usedMemory, entityCount, chunksLoaded));
            difference -= twentySeconds;
        }

        double tpsN = twentySeconds * 1.0 / difference;
```
The code matches the math.

----

That said, I found some oddity that removes 40ms from the difference before calculating the TPS, claiming to improve the accuracy & is likely the culprit.

It has been removed.

**Comment by Rsl1122** [comment]
Tests suggest that the inaccuracy has been fixed (/tps displays 19.98 and Plan displays 19.98)

**Comment by SlimeDog** [comment]
I agree. Testing verified correct TPS measurement in 4.1.6. Thank you.
